components/membership/membership-http-rpc-impl/src/main/kotlin/net/corda/membership/impl/httprpc/v1/MGMRpcOpsImpl.kt

```diff
@@ -0,0 +1,117 @@
+    private val mgmOpsClient: MGMOpsClient,
+    @Reference(service = MembershipGroupReaderProvider::class)
+    private val membershipGroupReaderProvider: MembershipGroupReaderProvider,
+    @Reference(service = VirtualNodeInfoReadService::class)
+    private val virtualNodeInfoReadService: VirtualNodeInfoReadService,
+) : MGMRpcOps, PluggableRPCOps<MGMRpcOps>, Lifecycle {
+    companion object {
+        private val logger: Logger = contextLogger()
+    }
+
+    private interface InnerMGMRpcOps {
+        fun generateGroupPolicy(holdingIdentityId: String): Set<Map.Entry<String, String?>>
+    }
+
+    private val className = this::class.java.simpleName
+
+    override val protocolVersion = 1
+
+    private var impl: InnerMGMRpcOps = InactiveImpl
+
+    private val coordinatorName = LifecycleCoordinatorName.forComponent<MGMRpcOps>(
+        protocolVersion.toString()
+    )
+
+    private val lifecycleHandler = RpcOpsLifecycleHandler(
+        ::activate,
+        ::deactivate,
+        setOf(LifecycleCoordinatorName.forComponent<MGMOpsClient>())
+    )
+
+    private val coordinator = coordinatorFactory.createCoordinator(coordinatorName, lifecycleHandler)
+
+    override val targetInterface: Class<MGMRpcOps> = MGMRpcOps::class.java
+
+    override val isRunning: Boolean
+        get() = coordinator.isRunning
+
+    override fun start() {
+        logger.info("$className started.")
+        coordinator.start()
+    }
+
+    override fun stop() {
+        logger.info("$className stopped.")
+        coordinator.stop()
+    }
+
+    override fun generateGroupPolicy(holdingIdentityId: String) =
+        impl.generateGroupPolicy(holdingIdentityId)
+
+    fun activate(reason: String) {
+        impl = ActiveImpl()
+        coordinator.updateStatus(LifecycleStatus.UP, reason)
+    }
+
+    fun deactivate(reason: String) {
+        coordinator.updateStatus(LifecycleStatus.DOWN, reason)
+        impl = InactiveImpl
+    }
+
+    private object InactiveImpl : InnerMGMRpcOps {
+        override fun generateGroupPolicy(holdingIdentityId: String) =
+            throw ServiceUnavailableException(
+                "${MGMRpcOpsImpl::class.java.simpleName} is not running. Operation cannot be fulfilled."
+            )
+    }
+
+    private inner class ActiveImpl : InnerMGMRpcOps {
+        override fun generateGroupPolicy(holdingIdentityId: String): Set<Map.Entry<String, String?>> {
+
+            val holdingIdentity = virtualNodeInfoReadService.getById(holdingIdentityId)?.holdingIdentity
+                ?: throw ResourceNotFoundException("Could not find holding identity associated with member.")
+
+            val reader = membershipGroupReaderProvider.getGroupReader(holdingIdentity)
+
+            val isMgm = reader.lookup(parse(holdingIdentity.x500Name))?.isMgm
+                ?: throw ResourceNotFoundException("The current member is not a MGM")
+
+            if(isMgm) {
+                val filteredMembers = reader.lookup(parse(holdingIdentity.x500Name))
```

Comment:
You're doing this lookup twice. Would be easier to just do it once. i.e. here and here: `val isMgm = reader.lookup(parse(holdingIdentity.x500Name))?.isMgm`

Reply:
Tried to value returned from lookup in `filteredMembers` outside the if block
and then tried to use `filteredMembers!!.isMgm` as the if condition but the compilation failed.
Also tried null check

Reply:
Perhaps something more kotlin-style would give better readability and avoid extra vals and lookups.
```suggestion
                val mgm = with (provider.getGroupReader(holdingIdentity)) {
                    lookup(parse(holdingIdentity.x500Name))?.apply {
                        require(isMgm) { "complain" }
                    } ?: throw Exception()
                }
```